### PR TITLE
ARROW-14676: [R][Docs] Add article on how to build a few different setups via docker to dev docs

### PR DESCRIPTION
**Pull request by thisisnic**
This PR will need rebasing from master after the PR for the debugging stuff is merged, but whilst working on this, it was simpler to create a branch from there so I can make the relevant `_pkgdown.yml` changes.

### CODE REVIEW

**Comment by thisisnic** [comment]
This needs some content on passing in variables via -e versus...the other way? Need to recall what I mean by this, think it's something to do with buildtime vs. runtime environments, though that may only be applicable to our CI jobs.

File: r/vignettes/developers/docker.Rmd [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+## How do I show all images saved?
+
+```shell
+docker images
+```
+
+## How do I show all running containers?
+
+```shell
+sudo docker ps
```
Comment (pitrou) [inline r/vignettes/developers/docker.Rmd:57]: Are you sure `sudo` is needed here?

Comment (thisisnic) [reply r/vignettes/developers/docker.Rmd:57]: It appears that is it not, thanks!
```suggestion
docker ps
```

File: r/vignettes/developers/docker.Rmd [hunk 2]
```diff
@@ -0,0 +1,71 @@
+# Using docker with Arrow
+
+Arrow is compatible with a huge number of combinations of OSs, OS versions, 
+compilers, R versions, and other variables.  Sometimes these combinations of
```
Comment (jonkeane) [inline r/vignettes/developers/docker.Rmd:4]: > a huge number of combinations of OSs, OS versions, compilers, R versions, and other variable

I feel like this is one of those lines that might be prefaced with "~unfortunately~ fortunately, ... " 😂 (not actually suggesting to add this, just commenting on the vibe)

File: r/vignettes/developers/docker.Rmd [hunk 3]
```diff
@@ -0,0 +1,71 @@
+# Using docker with Arrow
+
+Arrow is compatible with a huge number of combinations of OSs, OS versions, 
+compilers, R versions, and other variables.  Sometimes these combinations of 
+variables means that behaviours are found in some environments which cannot be 
+replicated in others.  In addition, there are different ways of building Arrow, 
+for example, using environment variables to specify the building of optional 
+components.
```
Comment (jonkeane) [inline r/vignettes/developers/docker.Rmd:8]:
```suggestion
for example, enabling the building of optional components.
```

Not sure we need to mention env vars here

**Review by jonkeane — COMMENTED** [review]
One concrete (small) suggestion. Generally speaking:

* do we want to add an example for how to run the docker-compose setup from the arrow directory? (e.g. `{tag vars} docker-compose run r`? This is what archery does, so maybe we 'just' add an example of that or add examples to those docs? I find that I do this type of build quite frequently (I don't think I've ever manually pulled from our docker hub cache, though maybe it would be quicker and I should!)
* We should probably mention something about needing to make sure that one is not using stale built artifacts which is the thrust of https://issues.apache.org/jira/browse/ARROW-14510 (or fix that somehow, `make clean` is probably not quite right, but the issue is that we don't want already-built binaries that happen to be in the /arrow/r directory to be used when we are running docker since those will be either broken or confusing — I believe basically all of the complaints of the docs build not working are related to this, and that build takes foooorever to build)
* Should we mention anything about crossbow here? I know we have it elsewhere, but it might be good to explain that crossbow uses this machinery as well

File: r/vignettes/developers/docker.Rmd [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+```shell
+docker run -it -e ARROW_DEPENDENCY_SOURCE=AUTO -v $(pwd):/arrow apache/arrow-dev:r-rhub-ubuntu-gcc-release-latest
+```
+
+Components:
+
+* `docker run` - command to run the container
+* `-it` - run with an interactive terminal so you can run commands on the containers
+* `-e ARROW_DEPENDENCY_SOURCE=AUTO` - set the environment variable `ARROW_DEPENDENCY_SOURCE` to the value `AUTO`
+* `-v $(pwd):/arrow` - mount the current directory at `/arrow` in the container
```
Comment (jonkeane) [inline r/vignettes/developers/docker.Rmd:33]:
```suggestion
* `-v $(pwd):/arrow` - mount the current directory at `/arrow` in the container (this command should be run in the root directory of a checkout of the arrow repo)
```

Comment (jonkeane) [reply r/vignettes/developers/docker.Rmd:33]: We also might want to put the warning about running this inside of `arrow/` up above (or both?)

**Review by jonkeane — APPROVED** [review]
This looks good, one small addition

**Pull request closed without merging**